src/k8s/pkg/k8sd/controllers/csrsigning/reconcile.go

```diff
@@ -0,0 +1,186 @@
+		}
+
+		derBytes, err := x509.CreateCertificate(rand.Reader, cert, caCert, certRequest.PublicKey, caKey)
+		if err != nil {
+			log.Error(err, "Failed to sign certificate")
+			return ctrl.Result{}, err
+		}
+		crtPEM = pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+		if crtPEM == nil {
+			log.Info("Failed to encode signed certificate to memory")
+			return ctrl.Result{RequeueAfter: requeueAfterSigningFailure}, nil
+		}
+	case "k8sd.io/kubelet-client":
+		caCert, caKey, err := pkiutil.LoadCertificate(config.Certificates.GetClientCACert(), config.Certificates.GetClientCAKey())
+		if err != nil {
+			log.Error(err, "Failed to load client CA certificate and key")
+			return ctrl.Result{}, err
+		}
+		cert := &x509.Certificate{
+			SerialNumber: serialNumber,
+			Subject: pkix.Name{
+				CommonName:   obj.Spec.Username,
+				Organization: obj.Spec.Groups,
+			},
+			NotBefore:             time.Now(),
+			NotAfter:              time.Now().AddDate(10, 0, 0), // TODO: expiration date from obj, or config
+			BasicConstraintsValid: true,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+			KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		}
+
+		derBytes, err := x509.CreateCertificate(rand.Reader, cert, caCert, certRequest.PublicKey, caKey)
+		if err != nil {
+			log.Error(err, "Failed to sign certificate")
+			return ctrl.Result{}, err
+		}
+		crtPEM = pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+		if crtPEM == nil {
+			log.Info("Failed to encode signed certificate to memory")
+			return ctrl.Result{RequeueAfter: requeueAfterSigningFailure}, nil
+		}
+	case "k8sd.io/kube-proxy-client":
+		caCert, caKey, err := pkiutil.LoadCertificate(config.Certificates.GetClientCACert(), config.Certificates.GetClientCAKey())
+		if err != nil {
+			log.Error(err, "Failed to load client CA certificate and key")
+			return ctrl.Result{}, err
+		}
+		cert := &x509.Certificate{
+			SerialNumber: serialNumber,
+			Subject: pkix.Name{
+				CommonName: "system:kube-proxy",
+			},
+			NotBefore:             time.Now(),
+			NotAfter:              time.Now().AddDate(10, 0, 0), // TODO: expiration date from obj, or config
+			BasicConstraintsValid: true,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+			KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		}
+
+		derBytes, err := x509.CreateCertificate(rand.Reader, cert, caCert, certRequest.PublicKey, caKey)
+		if err != nil {
+			log.Error(err, "Failed to sign certificate")
+			return ctrl.Result{}, err
+		}
+		crtPEM = pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+		if crtPEM == nil {
+			log.Info("Failed to encode signed certificate to memory")
+			return ctrl.Result{RequeueAfter: requeueAfterSigningFailure}, nil
+		}
+	default:
+		// NOTE(neoaggelos): this should never happen
+		return ctrl.Result{}, nil
```

Comment:
Would it be useful to log an error here?

Reply:
This really only comes in code because Go lacks proper sum types. Essentially, we have already limited the signerName values of this switch with https://github.com/canonical/k8s-snap/blob/3fcd43c2f7f7473cd75cb1fa8b18d7af353c812a/src/k8s/pkg/k8sd/controllers/csrsigning/setup.go#L31-L39

Perhaps the one thing to do here would be to log a warning, that "hey, things have really gone south here"